Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs

```diff
@@ -45,13 +72,108 @@ public abstract class BaseInputDeviceManager : BaseDataProvider<IMixedRealityInp
+        }
+
+        private PointerConfig[] pointerConfigurations = System.Array.Empty<PointerConfig>();
+
+        private class PointerEqualityComparer : IEqualityComparer<IMixedRealityPointer>
+        {
+            private static PointerEqualityComparer defaultComparer;
+
+            internal static PointerEqualityComparer Default
+            {
+                get =>  defaultComparer ?? (defaultComparer = new PointerEqualityComparer());
+            }
+
+            /// <summary>
+            /// Check that references equals for two pointers
+            /// </summary>
+            public bool Equals(IMixedRealityPointer p1, IMixedRealityPointer p2)
+            {
+                return ReferenceEquals(p1, p2);
+            }
+
+            /// <summary>
+            /// Unity objects have unique equals comparison and to check keys in a dictionary, 
+            /// we want the hascode match to be Unity's unique InstanceID to compare objects
+            /// </summary>
+            public int GetHashCode(IMixedRealityPointer pointer)
+            {
+                if (pointer is MonoBehaviour pointerObj)
+                {
+                    return pointerObj.GetInstanceID();
+                }
+                else
+                {
+                    return pointer.GetHashCode();
+                }
+            }
+        }
+
+        private static ProfilerMarker RequestPointersPerfMarker = new ProfilerMarker("Microsoft.MixedReality.Toolkit.Input.BaseInputDeviceManager.RequestPointers");
+
+        // Active pointers associated with the config index they were spawned from
+        private Dictionary<IMixedRealityPointer, uint> activePointersToConfig 
+            = new Dictionary<IMixedRealityPointer, uint>(PointerEqualityComparer.Default);
+
+        #endregion
+
+        #region IMixedRealityService implementation
 
         /// <inheritdoc />
-        public virtual IMixedRealityController[] GetActiveControllers() => System.Array.Empty<IMixedRealityController>();
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            if (InputSystemProfile != null && InputSystemProfile.PointerProfile != null)
+            {
+                var initPointerOptions = InputSystemProfile.PointerProfile.PointerOptions;
+
+                // If we were previously initialized, then clear our old pointer cache
+                if (pointerConfigurations != null && pointerConfigurations.Length > 0)
+                {
+                    DestroyPointerCache();
+                }
+
+                pointerConfigurations = new PointerConfig[initPointerOptions.Length];
+                activePointersToConfig.Clear();
+
+                for (int i = 0; i < initPointerOptions.Length; i++)
+                {
+                    pointerConfigurations[i].profile = initPointerOptions[i];
+                    pointerConfigurations[i].cache = new Stack<IMixedRealityPointer>();
```

Comment:
why is it a stack?

Reply:
Is this because you're thinking 'access the most recently lost item first' since hands tend to disappear and re-appear?

Reply:
yea just easy way to grab earliest item. I don't want a Pointer[] because of re-allocation and copying. I'm pretty sure List<> will also re-allocate and copy for when head is removed.  LinkedList or stack makes it easy to remove item